src/asn1.ml

```diff
@@ -615,11 +667,263 @@ struct
     map f g @@ sequence2
       (required ~label:"algorithm" Algo.grammar)
       (required ~label:"parameters" DH.Params.grammar)
+
 end
 
 let map_result f = function Result.Ok x -> Result.Ok (f x) | Result.Error _ as r -> r
 let default_result default = function Result.Error _ -> default () | Result.Ok _ as r -> r
 
+module CVC = struct
+  let rsa_oid = Asn.OID.of_string "0.4.0.127.0.7.2.2.2.1.4"
+  let ecdsa_oid = Asn.OID.of_string "0.4.0.127.0.7.2.2.2.2.3"
+
+  type t =
+    [ `RSA of RSA_CVC.Public.t | `ECDSA of ECDSA_CVC.Public.t | `UNKNOWN ]
+
+  type algo_typ =
+    | RSA
+    | ECDSA
+    | Unknown of Asn.OID.t
+
+  type parser_state =
+    | Init
+    | Type
+    | Length
+    | Value of int
+
+  let cvc_object_types =
+    [ 0x7F21, (`CV_CERTIFICATE, true)
+    ; 0x7F4E, (`CERTIFICATE_BODY, true)
+    ; 0x5F29, (`PROFILE_IDENTIFIER, false)
+    ; 0x7F49, (`PUBLIC_KEY, true)
+    ; 0x5F20, (`HOLDER_REFERENCE, false)
+    ; 0x7F4C, (`HOLDER_AUTH_TEMPLATE, true)
+    ; 0x5F25, (`EFFECTIVE_DATE, false)
+    ; 0x5F24, (`EXPIRATION_DATE, false)
+    ; 0x5F37, (`SIGNATURE, false)
+    ; 0x06, (`OID, false)
+    ; 0x42, (`CA_REFERENCE, false)
+    ; 0x67, (`REQ_AUTHENTICATION, true)
+    ; 0x53, (`ROLE_AND_ACCESS_RIGHTS, false)
+    ; 0x81, (`MODULUS, false)
+    ; 0x82, (`EXPONENT, false)
+    ; 0x82, (`COEFFICIENT_A, false)
+    ; 0x83, (`COEFFICIENT_B, false)
+    ; 0x84, (`BASE_POINT_G, false)
+    ; 0x85, (`BASE_POINT_R_ORDER, false)
+    ; 0x86, (`PUBLIC_POINT_Y, false)
+    ; 0x87, (`COFACTOR_F, false)
+    ]
+
+  let find_cvc_object_type tag =
+    let code =
+      Cstruct.get_char tag 0
+      |> Char.code
+    in
+    try code, List.assoc code cvc_object_types
+    with Not_found ->
+      let code =
+        let msb = code * 0x100 in
+        let lsb =
+          Cstruct.get_char tag 1
+          |> Char.code
+        in
+        msb + lsb
+      in
+      code, List.assoc code cvc_object_types
+
+  (* utility function to parse a big-endian blob as a Z.t *)
+  let atoz_bigendian s =
+    let reverse s =
+      let n = String.length s in
+      String.init n (fun i -> s.[n-1-i])
+    in
+    Z.of_bits @@ reverse @@ Cstruct.to_string s
+
+  let grammar =
+    let open Asn in
+    let f = function
+      | oid when oid = rsa_oid -> RSA
+      | oid when oid = ecdsa_oid -> ECDSA
+      | oid -> Unknown oid in
+    let g = function
+      | RSA -> rsa_oid
+      | ECDSA -> ecdsa_oid
+      | Unknown oid -> oid in
+    map f g oid
+
+  let decode_oid str =
+    let t, left = Asn.(decode_exn (codec ber grammar) str) in
```

Comment:
I couldn't find a specific helper for reporting leftover errors, so I used the method established elsewhere in the library.